GridListItem - horizontal alignment support

adids1221 opened the pull request:
## Description
Grid List Item support in horizontal alignment.
New `horizontalAlignment` prop to control the horizontal alignment of grid list item in all aspects - text, custom element, image and etc.

## Changelog
Grid List Item support in horizontal alignment

## Additional info
ticket 4082

Inbal-Tish commented on src/components/gridListItem/index.tsx:
```diff
@@ -1,13 +1,22 @@
+  right = 'right'
+}
+
+export type ItemHorizontalAlignmentProp = ItemHorizontalAlignment | `${ItemHorizontalAlignment}`;
```
This is not a good name...

Inbal-Tish replied:
Also, I don't see why we need to declare this type

Inbal-Tish commented on src/components/gridListItem/index.tsx:
```diff
@@ -85,6 +94,10 @@ export interface GridListItemProps {
+  /**
+   * Content horizontal alignment (default is center)
+   */
+  horizontalAlignment?: ItemHorizontalAlignmentProp;
```
Why do we need this prop? Why the user can't just pass 'alignItems' in the containerStyle?

adids1221 commented:
@Inbal-Tish 
I did few changes:
1.  Fixed the overlay use case to support the horizontal alignment
2. Changed the `children` type to solve ts error
3. Changed the prop name as we discussed
4. I spoke with Ethan for now we won't add any `textStyle` prop

Inbal-Tish commented on src/components/gridListItem/index.tsx:
```diff
@@ -101,7 +112,7 @@ export interface GridListItemProps {
    */
   testID?: string;
-  children?: React.ReactNode;
+  children?: React.ReactElement;
```
Can be an array

Inbal-Tish commented on src/components/image/index.tsx:
```diff
@@ -73,7 +73,7 @@ export type ImageProps = Omit<RNImageProps, 'source'> &
      * Render an overlay with custom content
      */
-    customOverlayContent?: JSX.Element;
+    customOverlayContent?: React.ReactElement;
```
Why the children type change is relevant to this PR?

Inbal-Tish commented on src/components/gridListItem/index.tsx:
```diff
@@ -139,14 +152,35 @@ class GridListItem extends Component<GridListItemProps> {
+            Typography[typography],
+            color && {color},
+            alignToStart && styles.contentAlignedToStart,
+            horizontalAlignmentStyle?.contentStyle
```
{textAlign: HorizontalAlignment} - can replace this object, just place it before this line: 
`alignToStart && styles.contentAlignedToStart,` to avoid overriding it

Inbal-Tish commented:
@adids1221 The code is very messy and unclear IMO. 
I would suggest to clean the code by:
1) in the Text pass:
`{textAlign: alignToStart ? 'left' : horizontalAlignment}`
2) in the container pass:
 `{alignItems: alignToStart ? 'flex-start' : this.getContainerHorizontalAlignment(horizontalAlignment)},`
3) make getContainerHorizontalAlignment function map to horizontalAlignment prop to alignItems values i.e. 'left' to 'flex-start'

Inbal-Tish commented on src/components/gridListItem/index.tsx:
```diff
@@ -181,33 +214,31 @@ class GridListItem extends Component<GridListItemProps> {
-        {hasOverlay && <View style={[styles.overlay, this.getItemSizeObj()]}>{renderOverlay?.()}</View>}
-        <TextContainer {...textContainerStyle}>
+        {renderOverlay && <View style={[styles.overlay, itemSize]}>{renderOverlay()}</View>}
+        <View {...textContainerStyle}>
```
You changed the TextContainer from View or Fragment to View. This can result in alignment issues (if you remember the Icon bug...). Why did you make this change?

Inbal-Tish commented on src/components/gridListItem/index.tsx:
```diff
@@ -181,33 +214,31 @@ class GridListItem extends Component<GridListItemProps> {
-        style={[styles.container, alignToStart && styles.containerAlignedToStart, {width}, containerStyle]}
+        style={[
+          styles.container,
+          {alignItems: alignToStart ? 'flex-start' : this.getContainerHorizontalAlignment(horizontalAlignment)},
```
I would move the function call outside of the style

Inbal-Tish commented on src/components/image/index.tsx:
```diff
@@ -73,7 +73,7 @@ export type ImageProps = Omit<RNImageProps, 'source'> &
      * Render an overlay with custom content
      */
-    customOverlayContent?: JSX.Element;
+    customOverlayContent?: React.ReactElement | React.ReactElement[];
```
What these changes have to do with this PR?
